ROVER-279 Funnels Errors from Supergraph Config Deserialisation to LSP

jonathanrainer opened the pull request:
**THIS CANNOT BE MERGED UNTIL ROVER-245 HAS**

So the key observation here is that the LSP needs to get back the details of errors that occur when we try and deserialise the Supergraph YAML and before this it didn't, those errors were simply printed into the DEBUG logs and then thrown away. In order to do this I've had to add a few more concepts into the work done on Subtasks etc. because in order to do this properly we need a Subtask that can send events across to multiple places (with broadcast semantics).

So the key changes in this PR follow the 3 commits:

1. Introduce the concept of a BroadcastSubtask, so this is a subtask which acts like an ordinary subtask, except that you can ask it for more streams of events and it will give them to you, via a `.subscribe()` method, which copies the interface of a `tokio` BroadcastStream. This does have consequences for the kind of data you can send out of these subtasks (i.e. all the data has to be Cloneable) but as this is a specific use-case I see no issue with this.
2. Introduce a new concept of a FederationWatcher (I'm not entirely sold on this name, and if someone has a better one let me know!) - This is essentially a Subtask that also watches the events produced by the SupergraphConfigWatcher but acts differently depending on what those events are. Initially it just printed but in the future we can use this new Subtask to watch, and hopefully update, the supergraph binary we're using (which is work that will come down the track in short order)
3. With all of that in mind, add the ability to feed those errors back into the LSP itself, so they display in the IDE, an example of which is shown below

![Screenshot 2025-01-08 at 13 37 58](https://github.com/user-attachments/assets/741f6b73-e099-4161-b636-730b9b20454f)

jonathanrainer committed:
ROVER-279 Introduce BroadcastSubtask

In order to route errors properly to the LSP we need to introduce
a new type of subtask which can broadcast out to multiple
receivers. This we have done, and have thus had to implement a few
changes to other underlying infrastructure, but not too much.

jonathanrainer committed:
ROVER-279 Introduce a Federation Watcher

This probably needs renaming but the idea is that this is
also watching events that come from watching the SupergraphConfig
and will also react to them.

jonathanrainer committed:
ROVER-279 Fix clippies and tests

aaronArinder commented on src/command/lsp/mod.rs:
```diff
@@ -0,0 +1,346 @@
+            FsWriteFile::default(),
+            client_config.service()?,
+            fetch_remote_subgraph_factory.boxed_clone(),
+            1,
```
worth being a const? what does this `1` represent, threads or something?

jonathanrainer replied:
So it's the retry interval for introspection, but this is actually fixed in the ROVER-245 PR because it gets turned into a command line option, which is a lot clearer anyway!

aaronArinder commented on src/composition/runner/mod.rs:
```diff
@@ -170,42 +175,73 @@ where
-            )
-        } else {
-            tracing::warn!(
+        let (supergraph_config_stream, supergraph_config_stream_2, supergraph_config_subtask) =
```
find `supergraph_config_stream_2` sort of unclear; any better names for it?

aaronArinder reviewed (approved):
this would have been a lot without your demo/walkthrough, so thanks; broadcastsubtask is pretty sweet!

jonathanrainer committed:
ROVER-279 Make variable names clearer

jonathanrainer committed:
ROVER-279 Add handling for invalid SupergraphConfig

The handling is rudimentary at the minute, but it surfaces errors
properly, which is the key
